File: Napakalaki/src/napakalaki/Napakalaki.java
/*
* PDOO Practicas - Napakalaki Java
* @author Emilio Reguero y Dunia Rodríguez 
*/
package napakalaki;

import java.util.ArrayList;
import java.util.Random;


public class Napakalaki {
    private static final Napakalaki instance = new Napakalaki();
    
    private Monster currentMonster;
    private Player currentPlayer;
    private ArrayList<Player> players = new ArrayList();
    private CardDealer dealer = CardDealer.getInstance();  
    
    private Napakalaki(){
    }
    
    private void initPlayers(ArrayList<String> names){
       
        for(String n : names){
           players.add(new Player(n));
        }
    }
    
    private Player nextPlayer(){
        int n;
        if(currentPlayer == null){
            Random rand = new Random();
            n = (rand.nextInt(players.size()));
        }else{
            n = players.indexOf(currentPlayer)+1;
            //0%2=0 1%2=1 2%2=0... 
            n = n % players.size();
        }
        currentPlayer = players.get(n);
        return currentPlayer;
    }
    
    private boolean nextTurnAllowed(){
        boolean allowed = false;
        if(currentPlayer == null || currentPlayer.validState())
           allowed=true;
            
        return allowed;
    }
    
    private void setEnemies(){
        
        Random rand = new Random();
        int pos = 0;
        
        for(Player p : players){
            int n = rand.nextInt(players.size());
            
            while(n == pos)
                n = rand.nextInt(players.size());
            
            p.setEnemy(players.get(n));
            pos = pos + 1;
        }
        
    }
    
    public static Napakalaki getInstance(){
        return instance;
    }
    
    public CombatResult developCombat(){ 
       CombatResult combatResult;
       combatResult = this.currentPlayer.combat(currentMonster);
       dealer.giveMonsterBack(currentMonster);
       if(combatResult == CombatResult.LOSEANDCONVERT){
           CardDealer cd = CardDealer.getInstance();
           Cultist c = cd.nextCultist();
           CultistPlayer cp = new CultistPlayer(currentPlayer,c);
           
           
           int pos = players.indexOf(currentPlayer);
           players.add(pos,cp);
           players.remove(pos+1);
           
           for(Player player : players){
               if(player != cp){
                   if(player.enemy == currentPlayer)
                       player.enemy = cp;
               }
           }
           
           currentPlayer = cp;
       }
        
       
       return combatResult;
    }
    
    public void discardVisibleTreasures (ArrayList<Treasure> treasures){
        for(Treasure treasure: treasures){
            currentPlayer.discardVisibleTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }
    
    public void discardHiddenTreasures (ArrayList<Treasure> treasures){
        for(Treasure treasure: treasures){
            currentPlayer.discardHiddenTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }
    
    public void makeTreasuresVisible(ArrayList<Treasure> treasures){
        for(Treasure t : treasures){
            currentPlayer.makeTreasureVisible(t);
        }
    }
    
    public void initGame(ArrayList<String> players){
        this.initPlayers(players);
        this.setEnemies();
        dealer.initCards();
        this.nextTurn();
    }
    
    public Player getCurrentPlayer(){
        return this.currentPlayer;
    }
    
    public Monster getCurrentMonster(){
        return currentMonster;
    }
    
    public boolean nextTurn(){
        boolean stateOK = nextTurnAllowed();
        if(stateOK){
            this.currentMonster = dealer.nextMonster();
            this.currentPlayer = this.nextPlayer();
            if(currentPlayer.isDead()){
                this.currentPlayer.initTreasures();
            }
        }
        
        return stateOK;
    }
    
    public boolean endOfGame(CombatResult result){
        boolean end = false;
        if(result == CombatResult.WINGAME)
            end = true;
        
        return end;
    }
}

File: NapakalakiRuby/pkg/NapakalakiRuby-0.0.1/lib/napakalaki.rb
#encoding: utf-8
#Autor: Dunia Rodríguez y Emilio Reguero
require_relative 'player'
require_relative 'card_dealer'
require_relative 'monster'
module NapakalakiGame
class Napakalaki
  
  include Singleton 
  
  attr_reader :currentPlayer, :currentMonster, :dealer, :players
  
  def initialize
    @players = Array.new
    @currentMonster = nil
    @currentPlayer = nil
    @dealer = CardDealer.instance
  end
  
  def getCurrentPlayer
    @currentPlayer
  end
  
  def getCurrentMonster
    @currentMonster
  end
 

  def developCombat()
    combat_result = @currentPlayer.combat(@currentMonster)
     if(combat_result ==  CombatResult::LOSEANDCONVERT)
       card = @dealer.next_cultist
       @players.delete(@currentPlayer)
       @currentPlayer = CultistPlayer.new(@currentPlayer, card)
       @players << @currentPlayer
     end

    @dealer.giveMonsterBack(@currentMonster)
    
    combat_result
  end
  
  def discardVisibleTreasures(treasures)
    for treasure in treasures
      @currentPlayer.discardVisibleTreasure(treasure)
      @dealer.giveTreasureBack(treasure)
    end
  end
  
  def discardHiddenTreasures(treasures)
     for treasure in treasures
      @currentPlayer.discardHiddenTreasure(treasure)
      @dealer.giveTreasureBack(treasure)
    end
  end
  
  def makeTreasuresVisible(treasures)
    i = 0
    while(i < treasures.size)
      t = treasures[i]
      @currentPlayer.makeTreasure_Visible(t)
      i = i +1  
    end
  end
  
  def initGame(players)
   initPlayers(players)
   setEnemies
   dealer.initCards
   nextTurn
  end
  
  def nextTurn()
    state_ok = nextTurnIsAllowed
    
    if state_ok 
      @currentMonster = dealer.nextMonster
      @currentPlayer = nextPlayer
      dead = currentPlayer.dead
      if dead
        @currentPlayer.initTreasures
      end
    end
    
    state_ok
  end
  
  def endOfGame(result)
    win_game = false
    if(result == CombatResult::WINGAME)
      win_game = true
    end
    
    win_game
  end
  
  
  private #A partir de aqui todos los metodos privados
  
  def initPlayers(names)
    names.each do |name|
      @players << Player.new(name)
    end
  end
  
  def nextPlayer()
    total_players = @players.length
    
    if(@currentPlayer == nil) then
      n = rand(total_players)
    else
      n = (@players.index(@currentPlayer))+1
      n = n % total_players;
    end
    
    next_player = @players.at(n)
    @currentPlayer = next_player
    
    @currentPlayer
  end
  
  def nextTurnIsAllowed()
    allowed = false
    if(@currentPlayer == nil || @currentPlayer.validState())
      allowed = true
    end
    
    allowed
  end
  
  def setEnemies()
    pos = 0
    @players.each do |player|
     
      n = rand(@players.length)
      while n == pos
         n = rand(@players.length)
      end
      
      player.setEnemy(@players[n])
    end
  end
  
   
  
end
end